pkg/controller/fileobserver/observer_linux.go

```diff
@@ -0,0 +1,116 @@
+// +build linux
+
+package fileobserver
+
+import (
+	"path/filepath"
+
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/golang/glog"
+	"github.com/sigma/go-inotify"
+)
+
+type linuxObserver struct {
+	watcher  *inotify.Watcher
+	queue    workqueue.Interface
+	reactors map[string][]reactorFn
+}
+
+func init() {
+	NewObserver = newLinuxObserver
+}
+
+func newLinuxObserver() (Observer, error) {
+	inotifyWatcher, err := inotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	return &linuxObserver{
+		watcher:  inotifyWatcher,
+		reactors: map[string][]reactorFn{},
+		queue:    workqueue.New(),
+	}, nil
+}
+
+func (o *linuxObserver) AddReactor(reactor reactorFn, filenames ...string) Observer {
+	for _, f := range filenames {
+		glog.V(5).Infof("Adding reactor for file %q", f)
+		o.watcher.AddWatch(filepath.Dir(f), inotify.IN_MODIFY|inotify.IN_CREATE|inotify.IN_DELETE)
```

Comment:
so this is watching the directory.  modifies come to the directory or to the file?

Reply:
file, the event always carry the file name (have test to prove that)